File: VB/WebSite/App_Code/NodeObjects.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Xpo

Public MustInherit Class NodeObject
	Inherits XPObject
	Public Sub New()
		MyBase.New()
	End Sub

	Public Sub New(ByVal session As Session)
		MyBase.New(session)
	End Sub

	Protected text_Renamed As String
	Public Property Text() As String
		Get
			Return text_Renamed
		End Get
		Set(ByVal value As String)
			SetPropertyValue(Of String)("Title", text_Renamed, value)
		End Set
	End Property
End Class

Public Class LevelOne
	Inherits NodeObject
	Public Sub New(ByVal session As Session)
		MyBase.New(session)
	End Sub

	<Association("Parent-Children")> _
	Public ReadOnly Property Children() As XPCollection(Of LevelTwo)
		Get
			Return GetCollection(Of LevelTwo)("Children")
		End Get
	End Property
End Class

Public Class LevelTwo
	Inherits NodeObject
	Public Sub New(ByVal session As Session)
		MyBase.New(session)
	End Sub

	Private _Parent As LevelOne
	<Association("Parent-Children")> _
	Public Property Parent() As LevelOne
		Get
			Return _Parent
		End Get
		Set(ByVal value As LevelOne)
			SetPropertyValue("Parent", _Parent, value)
		End Set
	End Property

	<Association("Owner-Details")> _
	Public ReadOnly Property Details() As XPCollection(Of LevelThree)
		Get
			Return GetCollection(Of LevelThree)("Details")
		End Get
	End Property
End Class

Public Class LevelThree
	Inherits NodeObject
	Public Sub New(ByVal session As Session)
		MyBase.New(session)
	End Sub

	Private _Owner As LevelTwo
	<Association("Owner-Details")> _
	Public Property Owner() As LevelTwo
		Get
			Return _Owner
		End Get
		Set(ByVal value As LevelTwo)
			SetPropertyValue("Owner", _Owner, value)
		End Set
	End Property
End Class

File: CS/WebSite/App_Code/NodeObjects.cs
using System;
using DevExpress.Xpo;

public abstract class NodeObject : XPObject {
    public NodeObject()
        : base() { }

    public NodeObject(Session session)
        : base(session) { }

    protected String text;
    public String Text {
        get { return text; }
        set { SetPropertyValue<String>("Title", ref text, value); }
    }
}

public class LevelOne : NodeObject {
    public LevelOne(Session session) : base(session) { }

    [Association("Parent-Children")]
    public XPCollection<LevelTwo> Children {
        get { return GetCollection<LevelTwo>("Children"); }
    }
}

public class LevelTwo : NodeObject {
    public LevelTwo(Session session) : base(session) { }

    LevelOne _Parent;
    [Association("Parent-Children")]
    public LevelOne Parent {
        get { return _Parent; }
        set { SetPropertyValue("Parent", ref _Parent, value); }
    }

    [Association("Owner-Details")]
    public XPCollection<LevelThree> Details {
        get { return GetCollection<LevelThree>("Details"); }
    }
}

public class LevelThree : NodeObject {
    public LevelThree(Session session) : base(session) { }

    LevelTwo _Owner;
    [Association("Owner-Details")]
    public LevelTwo Owner {
        get { return _Owner; }
        set { SetPropertyValue("Owner", ref _Owner, value); }
    }
}

